webhook/rollout/validating/rollout_create_update_handler.go

```diff
@@ -0,0 +1,279 @@
+	if canary == nil {
+		return field.ErrorList{field.Invalid(fldPath, nil, "Canary cannot be empty")}
+	}
+
+	errList := validateRolloutSpecCanarySteps(canary.Steps, fldPath.Child("Steps"))
+	errList = append(errList, validateRolloutSpecCanaryTraffic(canary.TrafficRouting, fldPath.Child("TrafficRouting"))...)
+	return errList
+}
+
+func validateRolloutSpecCanaryTraffic(traffic *appsv1alpha1.TrafficRouting, fldPath *field.Path) field.ErrorList {
+	if traffic == nil {
+		return field.ErrorList{field.Invalid(fldPath, nil, "Canary.TrafficRouting cannot be empty")}
+	}
+
+	errList := field.ErrorList{}
+	if len(traffic.Service) == 0 {
+		errList = append(errList, field.Invalid(fldPath.Child("Service"), traffic.Service, "TrafficRouting.Service cannot be empty"))
+	}
+
+	switch traffic.Type {
+	case "", appsv1alpha1.TrafficRoutingNginx:
+		if traffic.Nginx == nil ||
+			len(traffic.Nginx.Ingress) == 0 {
+			errList = append(errList, field.Invalid(fldPath.Child("Nginx"), traffic.Nginx, "TrafficRouting.Nginx.Ingress cannot be empty"))
+		}
+	default:
+
+		errList = append(errList, field.Invalid(fldPath.Child("Type"), traffic.Type, "TrafficRouting only support 'nginx' type"))
+	}
+
+	return errList
+}
+
+func validateRolloutSpecCanarySteps(steps []appsv1alpha1.CanaryStep, fldPath *field.Path) field.ErrorList {
+	stepCount := len(steps)
+	if stepCount == 0 {
+		return field.ErrorList{field.Invalid(fldPath, steps, "The number of Canary.Steps cannot be empty")}
+	}
+
+	if steps[stepCount-1].Weight != 100 {
+		return field.ErrorList{field.Invalid(fldPath, steps, "The 'Weight' field of the last CanaryStep must be '100'")}
+	}
+
+	for i := range steps {
+		s := &steps[i]
+		if s.Weight <= 0 || s.Weight > 100 {
+			return field.ErrorList{field.Invalid(fldPath.Index(i).Child("Weight"),
+				s.Weight, `Weight must be a positive number with "0" < weight <= "100"`)}
+		}
+		if s.Replicas != nil {
+			canaryReplicas, err := intstr.GetScaledValueFromIntOrPercent(s.Replicas, 100, true)
+			if err != nil || canaryReplicas <= 0 || canaryReplicas > 100 {
+				return field.ErrorList{field.Invalid(fldPath.Index(i).Child("CanaryReplicas"),
+					s.Replicas, `canaryReplicas must be positive number with with "0" < canaryReplicas <= "100", or a percentage with "0%" < canaryReplicas <= "100%"`)}
+			}
+		}
+	}
+
+	for i := 1; i < stepCount; i++ {
+		prev := &steps[i-1]
+		curr := &steps[i]
+		if curr.Weight < prev.Weight {
+			return field.ErrorList{field.Invalid(fldPath.Child("Weight"), steps, `Steps.Weight must be a non decreasing sequence`)}
+		}
+
+		// if they are comparable, then compare them
+		if IsPercentageCanaryReplicasType(prev.Replicas) != IsPercentageCanaryReplicasType(curr.Replicas) {
+			continue
+		}
+
+		prevCanaryReplicas, _ := intstr.GetScaledValueFromIntOrPercent(prev.Replicas, 100, true)
+		currCanaryReplicas, _ := intstr.GetScaledValueFromIntOrPercent(curr.Replicas, 100, true)
+		if prev.Replicas == nil {
+			prevCanaryReplicas = int(prev.Weight)
```

Comment:
weight is number between [1, 100], it is not comparable to replicas directly

Reply:
Line 234 filters out these un-comparable cases.